pkg/adapter/main.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"time"
+
+	// Uncomment the following line to load the gcp plugin
+	// (only required to authenticate against GKE clusters).
+	// _ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	cloudevents "github.com/cloudevents/sdk-go"
+	"github.com/kelseyhightower/envconfig"
+	"go.opencensus.io/stats/view"
+	"go.uber.org/zap"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/metrics"
+	"knative.dev/pkg/signals"
+	"knative.dev/pkg/source"
+
+	"knative.dev/eventing/pkg/kncloudevents"
+	"knative.dev/eventing/pkg/tracing"
+)
+
+type Adapter interface {
+	Start(stopCh <-chan struct{}) error
+}
+
+type AdapterConstructor func(ctx context.Context, env EnvConfigAccessor, client cloudevents.Client, reporter source.StatsReporter) Adapter
+
+func Main(component string, ector EnvConfigConstructor, ctor AdapterConstructor) {
+	flag.Parse()
+
+	ctx := signals.NewContext()
+
+	env := ector()
+	if err := envconfig.Process("", env); err != nil {
+		log.Fatalf("Error processing env var: %s", err)
+	}
+
+	// Convert json logging.Config to logging.Config.
+	loggingConfig, err := logging.JsonToLoggingConfig(env.GetLoggingConfigJson())
+	if err != nil {
+		fmt.Printf("[ERROR] failed to process logging config: %s", err.Error())
+		// Use default logging config.
+		if loggingConfig, err = logging.NewConfigFromMap(map[string]string{}); err != nil {
+			// If this fails, there is no recovering.
+			panic(err)
+		}
+	}
+
+	loggerSugared, _ := logging.NewLoggerFromConfig(loggingConfig, component)
+	logger := loggerSugared.Desugar()
+	defer flush(loggerSugared)
+	ctx = logging.WithLogger(ctx, loggerSugared)
```

Comment:
I realize that code is moved around, but I was wonder why bot suggared, and desuggard loggers.

How about smth like:

https://github.com/knative/eventing/blob/aca7b619ba28c8e9f85649ec413d636e3160503d/cmd/broker/ingress/main.go#L63-L64

Reply:
I changed to the same as this: https://github.com/knative/pkg/blob/master/injection/sharedmain/main.go#L137

Which makes me think the adapters should also watch for config changes. I'll open an issue for that.